### PR TITLE
Refactor error handling to allow use by custom post_send

### PR DESCRIPTION
**Pull request by alexmojaki**
Alternative to https://github.com/yola/demands/pull/43

Now this can be done instead:

    class SlackService(HTTPServiceClient):
    
        def post_send(self, response, **kwargs):
            if not response.json()['ok']:
                self.log_and_raise_error(response)
            return response

### CODE REVIEW

**Comment by RayeN** [comment]
I would say we are almost ok with #43 - so no need to for this one.

**Comment by alexmojaki** [comment]
If this was merged I would replace [this line](https://github.com/yola/asyncserver/pull/117/files#diff-e9a097a45affb9a770f84f35e045d77dR19) with this utility method. With #43 as it is I would instead do this:

    def call_method(self, api_method, **params):
        params['token'] = self.token
        params['channel'] = self.channel
        response = self.post(api_method, params=params)
        return DotDict(response.json())

    def is_success(self, response):
        return response.json()['ok']

I personally find the former solution is nicer: there's no double JSON parsing, I can use dot notation, there's fewer methods, and there's less burden on the reader to learn about the demands API: they don't have to know about hooks or distinction between `is_ok` and `is_success`.

**Comment by RayeN** [comment]
As I said before - there is no much sense in `log.error()` for asyncserver case, so you could just raise HTTPServiceError(response) in your `post_send()`.  So, don't see any benefit in this pull.

> If this was merged I would replace this line with this utility method.

Not that simple,  you'll have to additionally check for non-200 in your code

**Comment by RayeN** [comment]
This one is good, don't forget to fix tests.

**Comment by alexmojaki** [comment]
Firstly, I don't agree that logging is pointless just because something goes to Sentry. For example logging lets me tail a log file, and sentry doesn't work for local development.

Anyway, it's been pointed out that it's not good practice to both log an error and then raise it: you end up with too much in the logs. So I think we should move away from that and just make sure the message exists in the exception. It's already mostly there, but I see no reason for it to be incomplete. With this there is no unneeded logging, all the information will show up somewhere, the slack pull request is exactly as it needs to be, and everything is simple to understand. Also no I don't need to check for 200 in my code because demands still does that.

**Comment by RayeN** [comment]
> Firstly, I don't agree that logging is pointless just because something goes to Sentry. For example logging lets me tail a log file, and sentry doesn't work for local development.

Unhandled exception should go to log file as well.

> Anyway, it's been pointed out that it's not good practice to both log an error and then raise it: you end up with too much in the logs. So I think we should move away from that and just make sure the message exists in the exception. It's already mostly there, but I see no reason for it to be incomplete. With this there is no unneeded logging, all the information will show up somewhere, the slack pull request is exactly as it needs to be, and everything is simple to understand. Also no I don't need to check for 200 in my code because demands still does that.

Completely agree with that.  And sorry, missed the case with 200 check.

File: tests/test_demands.py [hunk 1]
```diff
@@ -148,17 +148,15 @@ def test_post_sends_no_exception_in_case_of_expected_response_code(self):
         self.service.request(
             'METHOD', 'http://notfound/', expected_response_codes=(404,))
 
-    @patch('demands.log')
-    def test_post_send_logs_errors(self, mock_log):
+    def test_post_send_logs_errors(self):
```
Comment (RayeN) [inline tests/test_demands.py]: we should either rename this test or remove it.  Because we don't log errors anymore.

File: demands/__init__.py [hunk 2]
```diff
@@ -22,7 +22,9 @@ def __init__(self, response):
         except ValueError:
             self.details = response.content
         super(AssertionError, self).__init__(
-            'code: %s, details: %s' % (response.status_code, self.details)
+            'Unexpected response from %s: url: %s, code: %s, details: %s' % (
+                self.__class__.__name__, response.url, response.status_code,
```
Comment (RayeN) [inline demands/__init__.py]: self.__class__.__name__ - make no sense right now.  Previously it was class name of certain service client, e.g. UserService, SlackService etc.  But rnow it's always `HTTPServiceError`.  I think we should just remove it.

File: demands/__init__.py [hunk 3]
```diff
@@ -15,14 +15,16 @@
...
         self.response = response
         try:
             self.details = response.json()
         except ValueError:
             self.details = response.content
         super(AssertionError, self).__init__(
-            'code: %s, details: %s' % (response.status_code, self.details)
+            'Unexpected response%s: url: %s, code: %s, details: %s' % (
+                ('from %s' % client.__class__.__name__) if client else '',
+                response.url, response.status_code, self.details)
```
Comment (RayeN) [inline demands/__init__.py]: I don't think client name is so important - because you have traceback - we can just omit it.

Comment (RayeN) [inline demands/__init__.py]: just 'Unexpected response: url: %s, code: %s, details: %s' would be probably enough.

Comment (alexmojaki) [inline demands/__init__.py]: You're right, the traceback is good enough

**Comment by RayeN** [comment]
:+1:

**Comment by alexmojaki** [comment]
I will bump the version after #43 is sorted out.

**Pull request merged**